pkg/scheduler/queue/tenant_queues.go

```diff
@@ -103,28 +142,46 @@ func (qb *queueBroker) makeQueuePath(request *tenantRequest) (QueuePath, error)
 	return QueuePath{string(request.tenantID)}, nil
 }
 
-func (qb *queueBroker) dequeueRequestForQuerier(lastTenantIndex int, querierID QuerierID) (*tenantRequest, *queueTenant, int, error) {
-	tenant, tenantIndex, err := qb.tenantQuerierAssignments.getNextTenantForQuerier(lastTenantIndex, querierID)
-	if tenant == nil || err != nil {
-		return nil, tenant, tenantIndex, err
+func (qb *queueBroker) dequeueRequestForQuerier(
+	lastTenantIndex int,
+	querierID QuerierID,
+) (
+	*tenantRequest,
+	*queueTenant,
+	int,
+	error,
+) {
+	// check if querier is registered and is not shutting down
+	if q := qb.tenantQuerierAssignments.queriersByID[querierID]; q == nil || q.shuttingDown {
+		return nil, nil, qb.tenantQuerierAssignments.tenantOrderIndex, ErrQuerierShuttingDown
 	}
 
-	queuePath := QueuePath{string(tenant.tenantID)}
-	queueElement := qb.tenantQueuesTree.DequeueByPath(queuePath)
+	qb.tenantQuerierAssignments.currentQuerier = &querierID
+	qb.tenantQuerierAssignments.tenantOrderIndex = lastTenantIndex
 
-	queueNodeAfterDequeue := qb.tenantQueuesTree.getNode(queuePath)
-	if queueNodeAfterDequeue == nil {
-		// queue node was deleted due to being empty after dequeue
-		qb.tenantQuerierAssignments.removeTenant(tenant.tenantID)
-	}
+	queuePath, queueElement := qb.tree.Dequeue()
```

Comment:
This is another place where state is mutated outside of the place it's used, which makes the whole thing harder to follow. e.g. when reading the code for the TQA, I was confused to see `currentQuerier` used but never set anywhere. Similarly, `tenantOrderIndex` is set both by the `queueBroker` and by the TQA, so I need to keep the behavior of both pieces in my head to verify it's doing what I expect.

Reply:
This shared-pointer approach is also the part of the design I would most want to avoid.

There's another approach I have thought about which has some aesthetic disadvantages but can address the shared-state concerns here.
Basically if you don't use a shared pointers you need another way to update provide external state updates before dequeuing - in this case, we need to provide the querier ID and lastTenantIndex every time, and then periodically need to update shuffle shards too.
When we add a least-load algorithm, we will also need to pass in that latest load data with the `QueryComponentUtilization` - this will replace the vanilla round-robin which doesn't need any external state.

Unfortunately it's impossible (I think) in Go to make any sort of common "update state" algorithm for these two different types implementing the dequeue interface.
So those two structs would each need a different "update state" "or prepare the dequeue ops" type method which we just have to make sure gets called by queueBroker before dequeues.
Because those methods can't have the same signature we can't do like a super cool generic thing that ignores them being different types, but I think the advantages might be worth it.

I can scratch it out on a branch here to demo if needed.